presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestAddNotNullFiltersToJoinNode.java

```diff
@@ -0,0 +1,321 @@
+    private final FunctionAndTypeManager functionAndTypeManager;
+
+    public TestAddNotNullFiltersToJoinNode()
+    {
+        super(ImmutableMap.of(OPTIMIZE_NULLS_IN_JOINS, Boolean.toString(false),
+                JOINS_NOT_NULL_INFERENCE_STRATEGY, USE_FUNCTION_METADATA.toString()));
+
+        functionAndTypeManager = createTestFunctionAndTypeManager();
+        Metadata metadata = MetadataManager.createTestMetadataManager();
+        rowExpressionTranslator = new TestingRowExpressionTranslator(metadata);
+    }
+
+    @DataProvider
+    public static Object[][] getExistingNotNullVarsTestCases()
+    {
+        return new Object[][] {
+                {"a IS NOT NULL AND b IS NOT NULL", new String[] {"a", "b"}},
+                {"a > 10 AND b IS NOT NULL AND c is NOT NULL", new String[] {"b", "c"}},
+                {"a is NULL AND b IS NOT NULL", new String[] {"b"}},
+                {"NOT(a is NULL)", new String[] {"a"}},
+                {"NOT(a is NULL OR b is NULL)", new String[] {}},
+                {"a is NOT NULL OR b is NOT NULL", new String[] {}}
+        };
+    }
+
+    @DataProvider
+    public static Object[][] nonStandardOperatorTestCases()
+    {
+        return new Object[][] {
+                {"NOT (b + 10 > c)", new String[] {"b", "c"}},
+                {"abs(b + c) > 10", new String[] {"b", "c"}},
+                {"random(b) = ceil(c)", new String[] {"b", "c"}},
+        };
+    }
+
+    @DataProvider
+    public static Object[][] standardOperatorTestCases()
+    {
+        return new Object[][] {
+                {"a + b > 10", new String[] {"a", "b"}},
+                {"a != 10 - b", new String[] {"a", "b"}},
+                {"a > b", new String[] {"a", "b"}},
+                {"a + NULL > b", new String[] {"a", "b"}},
+                // We can infer NOT NULL predicates on arguments of an AND expression
+                {"a > b and c = d", new String[] {"a", "b", "c", "d"}},
+                {"a IS NULL and b > c", new String[] {"b", "c"}},
+                // We cannot infer NOT NULL predicates on arguments of an OR expression
+                {"a > b OR c = d", new String[] {}},
+                // COALESCE can operate on NULL arguments, so cant infer predicates on its arguments
+                {"COALESCE(a,b)", new String[] {}},
+                // IN can operate on NULL arguments, so cant infer predicates on its arguments
+                {"a IN (b,10,NULL)", new String[] {}},
+                // arr[3] = 10 translates to EQUAL(SUBSCRIPT(arr, 3), 10). SUBSCRIPT is a standard Operator, so we can infer that 'arr' is NOT NULL
+                {"arr[3] = 10", new String[] {"arr"}},
+                // c_struct.a = 10 translates to EQUAL(DEREFERENCE(c_struct, 0), 10).
+                // We chose to not make any inferences for DEREFERENCE clauses, hence we don't add any NOT NULL clauses for 'c_struct' or 'c_struct.a'
+                {"c_struct.a = 10", new String[] {}},
```

Comment:
I've added these dereference test cases that illustrate how null inferencing would work for dereferenced variables. We will start by being conservative; this can be improved later. This addresses concerns raised in #16471 cc : @kaikalur

Reply:
Yes, the not null check will be added, but it does not cause any side-effects on query behavior.  
We observe the same failure modes as before  

Consider below table setups
```
presto:simple> create table t0 as select 10 as a0;
presto:simple> create table arrtest as select ARRAY[1, 2, 3] as arr3;
```


### With feature turned off `set session joins_not_null_inference_strategy='OFF';`
```
select 1 from t0 join arrtest on t0.a0 = arrtest.arr3[1]; //Runs fine
select 1 from t0 join arrtest on t0.a0 = arrtest.arr3[4]; //Fails with `Array subscript out of bounds`
```

### With feature turned on `set session joins_not_null_inference_strategy='USE_FUNCTION_METADATA';`
```
select 1 from t0 join arrtest on t0.a0 = arrtest.arr3[1]; //Runs fine. We add filterPredicates : `not(IS_NULL(a0))`, `not(IS_NULL(arr3[BIGINT'1']))`
select 1 from t0 join arrtest on t0.a0 = arrtest.arr3[4]; //Fails with `Array subscript out of bounds`. We add filterPredicates : `not(IS_NULL(a0))`, `not(IS_NULL(arr3[BIGINT'4]))`
```
As we see, the behavior with the feature off/on is unchanged.

Reply:
Also the reason we add the not null check on the dereference and not just the array is that the planner rewrites the derefrence to a new variablereference, so we see `com.facebook.presto.sql.planner.iterative.rule.AddNotNullFiltersToJoinNode	NotNull filters :: Existing : [], Inferred :[a0, expr_3]` in the debug log

Reply:
I cannot find the exact query which leads to the issue https://github.com/prestodb/presto/issues/15282, but one example to demonstrate what I mean here is `explain select * from t0 join arrtest on t0.a0 = arrtest.arr3[4] where cardinality(arr3) >= 4;` there will be filter "cardinality(arr3) >= 4 and not(isnull(arr3[4]))".

This query is OK here since `not(isnull(arr3[4]))` is after `cardinality(arr3) >= 4` and we will not evaluate `not(isnull(arr3[4]))` if `cardinality(arr3) >= 4` is false. But could be a problem if `not(isnull(arr3[4]))` is pushed below `cardinality(arr3) >= 4` or placed before it. I cannot find the exact query related to the issue. The previous approach is in query analysis phase, and this one is an optimizer, which can make a difference here.

I will let @kaikalur to chime in here as how to deal with this concern.

Reply:
It's not related to how to make "select * from arrtest where cardinality(arr3)>=4 and arr3[4]=1;" work. 

It's that, for query `select * from t0 join arrtest on t0.a0 = arrtest.arr3[4] where cardinality(arr3) >= 4;` which are guaranteed to work, there will be one more filter "not(isnull(arr3[4]))" added, and evaluation order of the filter expressions could lead to a problem.

Since I cannot find the query related to the mentioned issue to reproduce the problem, and the previous approach is in query analysis phase, and this one is an optimizer, which can make a difference here. I will leave it to you and other reviewers to decide whether to take care of this problem here.

One more thing, can you add the tests here in your PR too https://github.com/prestodb/presto/pull/16471/files#diff-68e1f723bdbf12a56f4a98f5212fa588c6d4ded0770068b465bfba336b55ef73 ?

Reply:
By the way, you can click the circle besides the reviewer name on the right top of page to "Re-request review".

Reply:
>It's that, for query select * from t0 join arrtest on t0.a0 = arrtest.arr3[4] where cardinality(arr3) >= 4; which are guaranteed to work, there will be one more filter "not(isnull(arr3[4]))" added, and evaluation order of the filter expressions could lead to a problem.

This query can be rewritten by the user to be more robust by : `select * from t0 join arrtest on t0.a0 = (select case when cardinality(arr3)>=4 then arr3[4] else NULL end);`. The NOT NULL constraints added for arrtest will be `filterPredicate = not(IS_NULL(SWITCH(BOOLEAN'true', WHEN((cardinality(arr3)) >= (BIGINT'4'), arr3[BIGINT'4']), null)))`

Reply:
>One more thing, can you add the tests here in your PR too [#16471 (files)](https://github.com/prestodb/presto/pull/16471/files#diff-68e1f723bdbf12a56f4a98f5212fa588c6d4ded0770068b465bfba336b55ef73) 

Yes, I will add these and re-ping

Reply:
@feilong-liu I [added](https://github.com/prestodb/presto/pull/18938/commits/982cb2130f1988d3104da2a29c32ba104558913e#diff-68e1f723bdbf12a56f4a98f5212fa588c6d4ded0770068b465bfba336b55ef73) the tests from #16471 as well